### ISSUE TITLE
Maybe types within object type

### ISSUE THREAD

**Issue by kure-** [facebook/flow#6135]
Hi there,
i'm kind of confused with flow behaviour over maybe types within object types. I'm about to build couple of configurable styled components in React, while I wish to set up some default values, so one doesn't have to fill all of the props in order to use the component. Some of the props can accept either a string or a number, e.g. for `width` which can then transform with glamorous by default to `px` when using number or simply use the string. 

[Example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoALgTwA4FMwDqAFgIbq4BuuATmALxgDO61AlgHYDmYAPmOwFcAtgCMaAblQYc+AArU42RvTABvHqgCQjOENzoiHTgC5CpclWoAaVDwC+k1AGM47ZmChw4KgBSqdegZGKgDkAIwATADMIXam8oqMAJT0AHxgLAK4ji5u6GAiJLQMPgH6hlymxGSUNKGRMSl06ZnZqEA)

As far as I've understood according to the example above, having maybe type is possible and works just fine when used outside of an object. But when I use the same type within object, it throws an error. I want to stick with `string | number` and avoid anything like `any`, which would work fine.

Is this working as intended or am I doing something wrong?:-)

Thank you

**Comment by kure-** [comment]
ping :+1:

**Comment by mrkev** [comment]
Ah, this is a weird syntax thing to do with default arguments. Let me explain real quick. 

Without types, your example is actually invalid JavaScript:

https://repl.it/@aykev/LastingSimpleDebugging-1

In that example, I commented out a line that fails. `foo2` is right though, and defines the behavior you want. This is because the syntax for destructuring and default argument assignment cannot be mixed in the intuitive way you tried to, but rather they have to be used in the way shown in `foo2`. In simpler terms, the default argument is the whole object.

This is how it looks with types on Flow:

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoALgTwA4FMwDqAFgIbq4BuuATmALxgDO61AlgHYDmYAPmOwFcAtgCMaAblQYc+AArU42RvTABvHqgCQjOENzoiHTgC5CpclWoAaVDwC+kqQGM47ZmChw4KgBSqdegZGdmCm8orKDP66+oZcxgAsdgCU9AB8YCwCuI5AA

The error is cryptic, but that's a separate issue. Thanks for pointing this out!
